### PR TITLE
[WIP] Add solutions for tests

### PR DESCRIPTION
**Pull request by joaquimg**
Fixes mocksolver to support querying solutions after solves.
Starts adding tests for solutions.

### CODE REVIEW

**Comment by joaquimg** [comment]
Aims to resolve #28

**Comment by joaquimg** [comment]
@odow I´d like to merge this before updating to the newest MOI

File: src/copy.jl [hunk 1]
```diff
@@ -1,8 +1,5 @@
-function MOI.copy!(dest::LinQuadOptimizer, src::MOI.ModelLike; copynames=false)
-    return MOIU.defaultcopy!(dest, src, copynames)
-end
-
-function MOI.copy!(dest::MOI.ModelLike,    src::LinQuadOptimizer; copynames=false)
+function MOI.copy!(dest::LinQuadOptimizer, src; copynames=false)
+    src::MOI.ModelLike
```
Comment (odow) [inline src/copy.jl]: What is the purpose of this? To reduce ambiguity? If so can we add a comment

File: src/mockoptimizer.jl [hunk 2]
```diff
@@ -216,21 +152,113 @@ mutable struct MockLinQuadOptimizer <: LQOI.LinQuadOptimizer
...
     end
-    return m
+    return instance
+end
+
+function unload(from,to,warn = true)
+    if !isempty(from)
+        out = from[1]
+        try
+            out = copy(from[1])
```
Comment (odow) [inline src/mockoptimizer.jl]: Why the try?

File: src/mockoptimizer.jl [hunk 3]
```diff
@@ -216,21 +152,113 @@ mutable struct MockLinQuadOptimizer <: LQOI.LinQuadOptimizer
...
-        m.params[string(name)] = value
-        setparam!(m.inner, string(name), value)
+        instance.params[string(name)] = value
+        setparam!(instance.inner, string(name), value)
     end
-    return m
+    return instance
+end
+
+function unload(from,to,warn = true)
```
Comment (odow) [inline src/mockoptimizer.jl]: Two things: docstring to say what it does, and spaces between arguments

File: src/mockoptimizer.jl [hunk 4]
```diff
@@ -216,21 +152,113 @@ mutable struct MockLinQuadOptimizer <: LQOI.LinQuadOptimizer
...
+    if !isempty(from)
+        out = from[1]
+        try
+            out = copy(from[1])
+        end
+        shift!(from)
+        return out
+    else
+        if warn
+            Compat.@warn("cant solve this model no extra solution")
```
Comment (odow) [inline src/mockoptimizer.jl]: Warnings and errors should be properly formatted English sentences. I don't know what that warning means.

File: src/mockoptimizer.jl [hunk 5]
```diff
@@ -244,12 +272,13 @@ function LQOI.get_number_quadratic_constraints(instance::MockLinQuadOptimizer)
...
-    m.l_rows = Int[]
-    m.q_rows = Int[]
-    for (name,value) in m.params
-        setparam!(m.inner, name, value)
+function MOI.empty!(instance::MockLinQuadOptimizer)
+    MOI.empty!(instance,nothing)
+    # do not empty param
+    instance.l_rows = Int[]
+    instance.q_rows = Int[]
+    for (name,value) in instance.params
```
Comment (odow) [inline src/mockoptimizer.jl]: Nit: space after comma (and above)

File: src/solver_interface.jl [hunk 6]
```diff
@@ -521,6 +521,17 @@ each linear constraint.
...
+"""
+get_farkas_dual_bounds!(m, x::Vector{Float64})
+
+Get the farkas dual (certificate of primal infeasiblility) for the variable
+boundss in the model `m`, and store in `x`. `x`must have one element for
+each variable.
+"""
+function get_farkas_dual_bounds!(m::LinQuadOptimizer, x)
+    nothing
+end
```
Comment (odow) [inline src/solver_interface.jl]: Woah. Can we make this a separate PR instead of bundling it with the testing change?

File: test/contconic.jl [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    LQOI.set_primal_status!(solver, MOI.FeasiblePoint)
+    LQOI.set_dual_status!(solver, MOI.FeasiblePoint)
+
+    LQOI.set_variable_primal_solution!(solver, [1.0, 0.0, 2.0])
+    LQOI.set_constraint_primal_solution!(solver, [1.0, 0.0, 2.0, 3.0, 2.0])
+    LQOI.set_quadratic_primal_solution!(solver, Float64[])
+
+    LQOI.set_variable_dual_solution!(solver, [0.0, 0.0, 0.0])
+    LQOI.set_constraint_dual_solution!(solver, [-0.0, 2.0, -0.0, -3.0, -1.0])
+    LQOI.set_quadratic_dual_solution!(solver, Float64[])
```
Comment (odow) [inline test/contconic.jl]: There must be a nice way to reduce this verbosity...

Comment (joaquimg) [reply test/contconic.jl]: do you mean the function name?

Comment (odow) [reply test/contconic.jl]: No something like
```julia
function set_solution(solver;
                      variable_dual_solution = Float64[],
                      quadratic_dual_solution = Float64[])
    LQOI.set_variable_dual_solution!(solver, variable_dual_solution)
    LQOI.set_quadratic_dual_solution!(solver, variable_dual_solution)
end

set_solution(solver;
             variable_dual_solution = [0.0, 0.0, 0.0])
```

Comment (joaquimg) [reply test/contconic.jl]: I’d be ok with something along those lines.
I’d go for:
set_solution(solver; variable_dual = [0.0, 0.0, 0.0])

**Comment by joaquimg** [comment]
@odow I added the new set solution style to intlinear.
Tell me if you agree.
If so I will do the same for the other tests and merge.

File: test/runtests.jl [hunk 8]
```diff
@@ -1,16 +1,19 @@
+push!(Base.LOAD_PATH,joinpath(dirname(@__FILE__),"..",".."))
```
Comment (odow) [inline test/runtests.jl]: That is this? Nit: spaces after `,`

File: src/mockoptimizer.jl [hunk 9]
```diff
@@ -216,21 +152,166 @@ mutable struct MockLinQuadOptimizer <: LQOI.LinQuadOptimizer
...
+end
+function set_primal_status!(instance::MockLinQuadOptimizer, input)
+    push!(instance.primal_status_stored, input)
+end
+function set_dual_status!(instance::MockLinQuadOptimizer, input)
+    push!(instance.dual_status_stored, input)
+end
+
+function set_solution!(instance::MockLinQuadOptimizer;
+    variable_primal = nothing,
```
Comment (odow) [inline src/mockoptimizer.jl]: These can be sensible defaults, e.g., `variable_primal = Float64[]`, then you can skip the `Nothing` checks.

File: src/mockoptimizer.jl [hunk 10]
```diff
@@ -216,21 +152,166 @@ mutable struct MockLinQuadOptimizer <: LQOI.LinQuadOptimizer
...
+    instance.inner.variable_primal_solution = unload(instance.variable_primal_solution_stored, instance.inner.variable_primal_solution)
+    instance.inner.variable_dual_solution = unload(instance.variable_dual_solution_stored, instance.inner.variable_dual_solution)
+
+    instance.inner.constraint_primal_solution = unload(instance.constraint_primal_solution_stored, instance.inner.constraint_primal_solution)
+    instance.inner.constraint_dual_solution = unload(instance.constraint_dual_solution_stored, instance.inner.constraint_dual_solution)
+
+    instance.inner.quadratic_primal_solution = unload(instance.quadratic_primal_solution_stored, instance.inner.quadratic_primal_solution)
+    instance.inner.quadratic_dual_solution = unload(instance.quadratic_dual_solution_stored, instance.inner.quadratic_dual_solution)
+
+    nothing
```
Comment (odow) [inline src/mockoptimizer.jl]: Style: `return` instead of `nothing`

File: test/intlinear.jl [hunk 11]
```diff
@@ -0,0 +1,70 @@
+function set_knapsacktest_solutions!(solver)
+    # SOLVE 1
+    LQOI.set_solution!(solver,
+    termination_status = MOI.Success,
```
Comment (odow) [inline test/intlinear.jl]: See comment about sensible defaults, then we can remove things like this.

(Also Style: indentation.)

File: test/runtests.jl (hunk 8 above)
Comment (joaquimg) [reply test/runtests.jl]: this is me forgetting to remove debug lines, I was waiting a bit more to remove since it was already there (and very helpful)

Comment (joaquimg) [reply test/runtests.jl]: with this you just include runtests

File: test/contconic.jl [hunk 12]
```diff
@@ -1,67 +1,49 @@
...
-    LQOI.set_constraint_primal_solution!(solver, [NaN, NaN])
-    LQOI.set_quadratic_primal_solution!(solver, Float64[])
-
-    LQOI.set_variable_dual_solution!(solver, [NaN])
-    LQOI.set_constraint_dual_solution!(solver, [1.0, -1.0])
-    LQOI.set_quadratic_dual_solution!(solver, Float64[])
-
-    nothing
+    LQOI.set_solution!(solver,
+        termination_status = MOI.Success,
```
Comment (odow) [inline test/contconic.jl]: Success is the default

File: src/mockoptimizer.jl [hunk 13]
```diff
@@ -240,7 +240,7 @@ function fakesolve(instance::MockLinQuadOptimizer)
     instance.inner.quadratic_primal_solution = unload(instance.quadratic_primal_solution_stored, instance.inner.quadratic_primal_solution)
     instance.inner.quadratic_dual_solution = unload(instance.quadratic_dual_solution_stored, instance.inner.quadratic_dual_solution)
 
-    nothing
+    return nothing
```
Comment (odow) [inline src/mockoptimizer.jl]: Style nit: just `return` instead of `return nothing`

File: test/contquadratic.jl [hunk 14]
```diff
@@ -1,144 +1,125 @@
...
-
-    LQOI.set_variable_dual_solution!(solver, [0.0, 0.0, 0.0])
-    LQOI.set_constraint_dual_solution!(solver, [0.714286, 0.857144])
-    LQOI.set_quadratic_dual_solution!(solver, Float64[])
-
-    nothing
+    LQOI.set_solution!(solver,
+        termination_status = MOI.Success,
+        primal_status = MOI.FeasiblePoint,
+        dual_status = MOI.FeasiblePoint,
```
Comment (odow) [inline test/contquadratic.jl]: Success and FeasiblePoint are defaults. There are lots of other tests with these and similar defaults that don't need to be given.

File: src/mockoptimizer.jl (hunk 13 above)
Comment (joaquimg) [reply src/mockoptimizer.jl]: where is that in the style guide?
I don't like `return` alone...

**Comment by odow** [comment]
I just restarted the PR build on travis to check against the latest master now that #37 was merged. Providing it passes, I'll merge.
